--- a/source/lib/sysdep/arch.h
+++ b/source/lib/sysdep/arch.h
@@ -85,0 +86,6 @@
+#if defined(__wasm64__)
+# define ARCH_WASM64 1
+#else
+# define ARCH_WASM64 0
+#endif
+
@@ -87 +93 @@
-#if (ARCH_IA32+ARCH_IA64+ARCH_AMD64+ARCH_ALPHA+ARCH_ARM+ARCH_AARCH64+ARCH_MIPS+ARCH_E2K+ARCH_PPC64) != 1
+#if (ARCH_IA32+ARCH_IA64+ARCH_AMD64+ARCH_ALPHA+ARCH_ARM+ARCH_AARCH64+ARCH_MIPS+ARCH_E2K+ARCH_PPC64+ARCH_WASM64) != 1
