diff --git a/src/job.c b/src/job.c
index acb5317..a3896cf 100644
--- a/src/job.c
+++ b/src/job.c
@@ -76,7 +76,11 @@ char * vms_strsignal (int status);
 
 #else
 
+#ifdef __dolly__
+const char *default_shell = "/bin/slop";
+#else
 const char *default_shell = "/bin/sh";
+#endif
 int batch_mode_shell = 0;
 
 #endif
@@ -2057,6 +2061,8 @@ load_too_high (void)
 {
-#if defined(__MSDOS__) || defined(VMS) || defined(_AMIGA) || defined(__riscos__)
+#ifdef __dolly__
+  return 0;
+#elif defined(__MSDOS__) || defined(VMS) || defined(_AMIGA) || defined(__riscos__)
   return 1;
 #else
   static double last_sec;
   static time_t last_now;
diff --git a/src/main.c b/src/main.c
index 3aa975b..ad83f12 100644
--- a/src/main.c
+++ b/src/main.c
@@ -1164,14 +1164,17 @@ temp_stdin_unlink ()
 extern char **environ;
 #endif
 
-#if defined(_AMIGA) || defined(MK_OS_ZOS)
+#if defined(_AMIGA) || defined(MK_OS_ZOS) || defined(__dolly__)
 int
 main (int argc, char **argv)
 #else
 int
 main (int argc, char **argv, char **envp)
 #endif
 {
+#ifdef __dolly__
+  char **envp = environ;
+#endif
   int makefile_status = MAKE_SUCCESS;
   struct goaldep *read_files;
   PATH_VAR (current_directory);
diff --git a/src/commands.h b/src/commands.h
--- a/src/commands.h
+++ b/src/commands.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/config.h b/src/config.h
--- a/src/config.h
+++ b/src/config.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/debug.h b/src/debug.h
--- a/src/debug.h
+++ b/src/debug.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/dep.h b/src/dep.h
--- a/src/dep.h
+++ b/src/dep.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/filedef.h b/src/filedef.h
--- a/src/filedef.h
+++ b/src/filedef.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/gmk-default.h b/src/gmk-default.h
--- a/src/gmk-default.h
+++ b/src/gmk-default.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/hash.h b/src/hash.h
--- a/src/hash.h
+++ b/src/hash.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/job.h b/src/job.h
--- a/src/job.h
+++ b/src/job.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/makeint.h b/src/makeint.h
--- a/src/makeint.h
+++ b/src/makeint.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/os.h b/src/os.h
--- a/src/os.h
+++ b/src/os.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/output.h b/src/output.h
--- a/src/output.h
+++ b/src/output.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/rule.h b/src/rule.h
--- a/src/rule.h
+++ b/src/rule.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/shuffle.h b/src/shuffle.h
--- a/src/shuffle.h
+++ b/src/shuffle.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/src/variable.h b/src/variable.h
--- a/src/variable.h
+++ b/src/variable.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/lib/fnmatch.h b/lib/fnmatch.h
--- a/lib/fnmatch.h
+++ b/lib/fnmatch.h
@@ -0,0 +1 @@
+#pragma once
diff --git a/lib/glob.h b/lib/glob.h
--- a/lib/glob.h
+++ b/lib/glob.h
@@ -0,0 +1 @@
+#pragma once
