--- a/src/unistd.rs
+++ b/src/unistd.rs
@@ -1292,6 +1292,7 @@
 /// will be returned if the name is not valid or the current process does not
 /// have permissions to update the host name.
 #[cfg(not(target_os = "redox"))]
+#[cfg(not(target_os = "emscripten"))]
 pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
     // Handle some differences in type of the len arg across platforms.
     cfg_if! {
