--- a/Cargo.toml
+++ b/Cargo.toml
@@ -897,7 +897,7 @@
 [target.'cfg(all(tokio_unstable, target_os = "linux"))'.dev-dependencies.backtrace]
 version = "0.3.58"
 
-[target.'cfg(any(not(target_family = "wasm"), all(target_os = "wasi", not(target_env = "p1"))))'.dependencies.socket2]
+[target.'cfg(any(not(target_family = "wasm"), all(target_os = "emscripten", target_arch = "wasm64"), all(target_os = "wasi", not(target_env = "p1"))))'.dependencies.socket2]
 version = "0.6.3"
 features = ["all"]
 optional = true
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -466,6 +466,7 @@
 #[cfg(all(
     not(tokio_unstable),
     target_family = "wasm",
+    not(all(target_os = "emscripten", target_arch = "wasm64")),
     any(
         feature = "fs",
         feature = "io-std",
--- a/src/net/unix/ucred.rs
+++ b/src/net/unix/ucred.rs
@@ -1,4 +1,9 @@
 use crate::net::unix;
+
+#[cfg(all(target_os = "emscripten", target_arch = "wasm64"))]
+pub(crate) fn get_peer_cred(_: &unix::UnixStream) -> std::io::Result<UCred> {
+    Err(std::io::Error::from_raw_os_error(libc::ENOSYS))
+}
 
 /// Credentials of a process.
 #[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
